src/main/java/com/tabnine/selections/AutoImporter.java

```diff
@@ -32,197 +31,201 @@
+            ((EditorEx) editor)
+                    .getFilteredDocumentMarkupModel()
+                    .addMarkupModelListener(myDisposable, this);
+        }
+    }
+
+    public static void registerTabNineAutoImporter(@NotNull Editor editor, @NotNull Project project, int startOffset, int endOffset) {
+        AutoImporter autoImporter = editor.getUserData(TABNINE_AUTO_IMPORTER_KEY);
+        if (autoImporter != null) {
+            autoImporter.cleanup();
+            editor.putUserData(TABNINE_AUTO_IMPORTER_KEY, null);
+        }
+        autoImporter = new AutoImporter(editor, project, startOffset, endOffset);
+        editor.putUserData(TABNINE_AUTO_IMPORTER_KEY, autoImporter);
+        autoImporter.init();
+    }
+
+    private boolean isPriorHighlighter(@NotNull RangeHighlighterEx highlighter) {
+        return startOffset > highlighter.getAffectedAreaStartOffset();
+    }
+
+    private boolean isPosteriorHighlighter(@NotNull RangeHighlighterEx highlighter) {
+        return endOffset < highlighter.getAffectedAreaEndOffset();
+    }
+
+    @Nullable
+    private PsiFile getFile(@NotNull RangeHighlighterEx highlighter) {
+        final Document document = highlighter.getDocument();
+        return PsiDocumentManager.getInstance(project).getPsiFile(document);
+    }
+
+    @Nullable
+    private HighlightInfo getHighlightInfo(@NotNull RangeHighlighterEx highlighter) {
+        Object errorTooltip = highlighter.getErrorStripeTooltip();
+        return ObjectUtils.tryCast(errorTooltip, HighlightInfo.class);
+    }
+
+    private void invokeLater(@NotNull Runnable task) {
+        ApplicationManager.getApplication().invokeLater(task);
+    }
+
+    private void autoImportUsingCodeAnalyzer(@NotNull final PsiFile file) {
+        if (codeAnalyzerAutoImportInvoked.get()) {
+            return;
+        }
+        final DaemonCodeAnalyzerImpl codeAnalyzer =
+                (DaemonCodeAnalyzerImpl) DaemonCodeAnalyzer.getInstance(project);
+        CommandProcessor.getInstance()
+                .runUndoTransparentAction(() -> invokeLater(() -> {
+                    codeAnalyzerAutoImportInvoked.set(true);
+                    codeAnalyzer.autoImportReferenceAtCursor(editor, file);
+                }));
+    }
+
+    private void collectImportFixes(
+            @NotNull PsiFile file,
+            @NotNull RangeHighlighterEx highlighter,
+            @NotNull String highlightedText) {
+        List<HighlightInfo.IntentionActionDescriptor> availableFixes =
+                ShowIntentionsPass.getAvailableFixes(editor, file, -1, highlighter.getEndOffset());
+        List<HighlightInfo.IntentionActionDescriptor> importActions =
+                availableFixes.stream()
+                        .filter(f -> {
+                            String actionText = f.getAction().getText().toLowerCase();
+                            return actionText.contains("import") && !actionText.contains("remove");
```

Comment:
the entire file has been formatted, that's why the entire file is being shown at the diff. This line is the only relevant change.